Core/clim-core/panes/layout-protocol.lisp

```diff
@@ -466,20 +454,35 @@ which changed during the current execution of CHANGING-SPACE-REQUIREMENTS.
                                               &rest space-req-keys
                                               &key resize-frame &allow-other-keys)
   (declare (ignore resize-frame space-req-keys))
-  ;; Clear current width and height.
-  (setf (pane-space-requirement pane) nil
-        (pane-current-width pane) nil
-        (pane-current-height pane) nil))
+  ;; Clear the cached value
+  (setf (pane-space-requirement pane) nil))
 
 (defmethod change-space-requirements ((pane layout-protocol-mixin)
                                       &key resize-frame &allow-other-keys)
-  (when-let ((parent (sheet-parent pane)))
-    (change-space-requirements parent :resize-frame resize-frame)))
+  ;; do nothing here
+  nil)
 
 (defmethod change-space-requirements :after ((pane layout-protocol-mixin)
                                              &key resize-frame &allow-other-keys)
-  (declare (ignore resize-frame))
-  (note-space-requirements-changed (sheet-parent pane) pane))
+  (when-let ((parent (sheet-parent pane)))
+    (if resize-frame
+        ;; From Spec 29.3.4: "If resize-frame is true, then
+        ;; layout-frame will be invoked on the frame". Here instead of
+        ;; call directly LAYOUT-FRAME, we call
+        ;; CHANGE-SPACE-REQUIREMENTS on the parent and it travels all
+        ;; the way up to the top-level-sheet-pane which then invokes
+        ;; the layout protocol calling LAYOUT-FRAME. The rationale of
+        ;; this is:
+        ;; 1. we can't call (LAYOUT-FRAME (PANE-FRAME pane)) on a
+        ;;   menu because with the actual implementation of menu it
+        ;;   will layout the main application and not the menu frame.
+        ;; 2. we automatically clear the cached values of
+        ;;    space-requirements for the involved panes.
```

Comment:
do I understand correctly, that previously they were not invalidated and allocate-space was using obsolete results of compute-space?

assuming that we call layout-frame (i.e because the menu implementation is fixed to work better), how should that work? layout-protocol should invalidate all cached values at one go?  or should it be a responsibility of note-sheet-region-changed?

Reply:
We have to think about that, maybe it is necessary to change when cache the values and when invalidate them. I wrote something in the comment to your gist about layout-protocol https://gist.github.com/dkochmanski/d1dcacabcfa35886000301a8882f0b2c